### PR TITLE
importccl: add support for SQLFns in pgdump

### PR DESCRIPTION
**Pull request by madelynnblue**
SQLFns are functions users can call from SQL that create and execute
SQL. AddGeometryColumn, for example, creates an ADD COLUMN statement and
executes it. Add support for these in IMPORT PGDUMP by detecting them,
running them, and feeding them back in to the table statement reader.
    
Release note (sql change): add support for AddGeometryColumn and other
functions that mutate schema in IMPORT PGDUMP.

### CODE REVIEW

**Comment by mjibson** [comment]
RFAL @pbardea @Anzoteh96

**Review by dt — COMMENTED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50850)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Anzoteh96, @dt, @mjibson, @otan, and @pbardea)

___
*[pkg/ccl/importccl/read_import_pgdump.go, line 400 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50850#-MBF2plk7sZYcUbQPSZy:-MBF2plk7sZYcUbQPSZz:b-47s5ps) ([raw file](https://github.com/cockroachdb/cockroach/blob/5990465c1056aab842bcbc1d41bc64f570b06a8f/pkg/ccl/importccl/read_import_pgdump.go#L400)):*
> ```Go
> 	case *tree.Select:
> 		switch sel := stmt.Select.(type) {
> 		case *tree.SelectClause:
> ```

should this have a `default` counterpart for unsupported?
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50850#-MBF2ubd0WdxaRVEkMla:-MBF2ubd0WdxaRVEkMlb:bzhs08v) ([raw file](https://github.com/cockroachdb/cockroach/blob/5990465c1056aab842bcbc1d41bc64f570b06a8f/pkg/ccl/importccl/read_import_pgdump.go#L403)):*
> ```Go
> 			for _, selExpr := range sel.Exprs {
> 				switch expr := selExpr.Expr.(type) {
> 				case *tree.FuncExpr:
> ```

ditto, what if it is something else?
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 414 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50850#-MBEsxTl5oPWSF7uIdKW:-MBEsxTl5oPWSF7uIdKX:bu26yr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5990465c1056aab842bcbc1d41bc64f570b06a8f/pkg/ccl/importccl/read_import_pgdump.go#L414)):*
> ```Go
> 					fn := ov.SQLFn
> 					if fn == nil {
> 						continue
> ```

huh, in what cases do we get here / why are they okay to ignore?

**Review by madelynnblue — COMMENTED** [review]
I've added some comments regarding what to do about non schema-transforming SELECTs, and how we just ignore them. This was previously the implicit policy, and I'm making it more explicit with these comments. Since (AFAIK) no users have complained about IMPORT PGDUMP not executing their SELECTs (except in this geospatial case), I think it's ok to ignore them. Thoughts?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50850#-:-MBFKQYK05QSzjh5z9lY:b-qqumx1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Anzoteh96, @dt, @otan, and @pbardea)

___
*[pkg/ccl/importccl/read_import_pgdump.go, line 400 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50850#-MBF2plk7sZYcUbQPSZy:-MBFJIYG-gi0ROb8sivy:b-p92nbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5990465c1056aab842bcbc1d41bc64f570b06a8f/pkg/ccl/importccl/read_import_pgdump.go#L400)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

should this have a `default` counterpart for unsupported?
</blockquote></details>

I don't know? We were already ignoring any SELECTs, so I don't see why we would disallow them in this commit.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50850#-MBF2ubd0WdxaRVEkMla:-MBFJbaXBeC11KnS7GzP:b-4sbyi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5990465c1056aab842bcbc1d41bc64f570b06a8f/pkg/ccl/importccl/read_import_pgdump.go#L403)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

ditto, what if it is something else?
</blockquote></details>

Same as above: we already were ignoring these. If it is decided to care about erroring on SELECTs on other unsupported things, those should be added and tested in another PR.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 414 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50850#-MBEsxTl5oPWSF7uIdKW:-MBFJw_D1fvwYnPIxUO6:b-ndfi87) ([raw file](https://github.com/cockroachdb/cockroach/blob/5990465c1056aab842bcbc1d41bc64f570b06a8f/pkg/ccl/importccl/read_import_pgdump.go#L414)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

huh, in what cases do we get here / why are they okay to ignore?
</blockquote></details>

A non-sqlfn is ignored just like they were before this pr.

File: pkg/ccl/importccl/read_import_pgdump.go [hunk 1]
```diff
@@ -299,83 +299,161 @@ func readPostgresCreateTable(
...
+		switch sel := stmt.Select.(type) {
+		case *tree.SelectClause:
+			for _, selExpr := range sel.Exprs {
+				switch expr := selExpr.Expr.(type) {
+				case *tree.FuncExpr:
+					// Look for function calls that mutate schema (this is actually a thing).
+					semaCtx := tree.MakeSemaContext()
+					_, err := expr.TypeCheck(ctx, &semaCtx, nil /* desired */)
+					if err != nil {
+						return err
```
Comment (Anzoteh96) [inline pkg/ccl/importccl/read_import_pgdump.go:410]: Personal preference (take with a grain of salt :) ), I would condense this to 
`
if _, err := expr.TypeCheck(ctx, &semaCtx, nil); err != nil {
  return err
}
`

Comment (mjibson) [reply pkg/ccl/importccl/read_import_pgdump.go:409]: done

**Review by dt — APPROVED** [review]

**Pull request merged**